README.md
# hart-protocol

[![PyPI](https://img.shields.io/pypi/v/hart-protocol)](https://pypi.org/project/hart-protocol)
[![Conda](https://img.shields.io/conda/vn/conda-forge/hart-protocol)](https://anaconda.org/conda-forge/hart-protocol)
[![black](https://img.shields.io/badge/code--style-black-black)](https://black.readthedocs.io/)
[![ver](https://img.shields.io/badge/calver-YYYY.M.MICRO-blue)](https://calver.org/)
[![log](https://img.shields.io/badge/change-log-informational)](https://github.com/yaq-project/hart-protocol/-/blob/main/CHANGELOG.md)

A sans I/O Python implementation of the [Highway Adressable Remote Transducer Protocol](https://en.wikipedia.org/wiki/Highway_Addressable_Remote_Transducer_Protocol).

## Introduction

This Python package contains tooling for encoding and decoding bytestrings for communication with HART peripherals.
HART has been implemented using a variety of transport layers---Bell 202, RS485, Ethernet, etc.
In persuit of simplicity and reusability, this package does not contain any interface capabilities.
Use something like [pySerial](https://pyserial.readthedocs.io) for transport.
Read the [sans I/O manifesto](https://sans-io.readthedocs.io/) for more motivation regarding this design pattern.

Briefly, HART is an open protocol for industrial automation supported by multiple device manufacturers.
HART has a concept of "address", so that many peripherals can share the same communication channel.
HART has limited support for multiple controllers, and generic handheld controllers exist.
HART peripherals respond to numbered commands, which can be thought of as primative [remote procedure calls](https://en.wikipedia.org/wiki/Remote_procedure_call).
The standard specifies a number of universal commands which should be supported by any peripheral, and there are also so-called "common" commands which many peripherals implement.
It's strongly recommended that you check the documentation of your own peripheral---implementations may be inconsistent.
In addition to universal and common commands, it's likely that your peripheral implements many device-specific commands.

This package aims to have complete and accurate support for all universal and common commands.
In addition, this package has tooling for packing and unpacking generic command data for device-specific commands.
This package is intentionally simple and narrowly scoped.
There is no documentation beyond this README.
Please open an issue or PR to the GitHub repository if you find any errors or missing functionality.

## Sending Commands

The following functions return bytestrings that can be fed to your transport layer.

Universal Commands
| command | function                                                    |
| ------- | ----------------------------------------------------------- |
| 0       | `read_unique_identifier(address)`                           |
| 1       | `read_primary_variable(address)`                            |
| 2       | `read_loop_current_and_percent(address)`                    |
| 3       | `read_dynamic_variables_and_loop_current(address)`          |
| 6       | `write_polling_address(address, new_short_address)`         |
| 11      | `read_unique_identifier_associated_with_tag(tag)`           |
| 12      | `read_message(address)`                                     |
| 13      | `read_tag_descriptor_date(address)`                         |
| 14      | `read_primary_variable_information(address)`                |
| 15      | `read_output_information(address)`                          |
| 16      | `read_final_assembly_number(address)`                       |
| 17      | `write_message(address, message)`                           |
| 18      | `write_tag_descriptor_date(address, tag, descriptor, date)` |
| 19      | `write_final_assembly_number(address, number)`              |

Common-Practice Commands
| command | function                                                 |
| ------- | -------------------------------------------------------- |
| 37      | `set_primary_variable_lower_range_value(address, value)` |
| 38      | `reset_configuration_changed_flag(address)`              |
| 42      | `perform_master_reset(address)`                          |
| 48      | `read_additional_transmitter_status(address)`            |
| 50      | `read_dynamic_variable_assignments(address)`             |
| 59      | `write_number_of_response_preambles(address, number)`    |
| 66      | `toggle_analog_output_mode(address)`                     |
| 67      | `trim_analog_output_zero(address)`                       |
| 68      | `trim_analog_output_span(address)`                       |
| 123     | `select_baud_rate(address, rate)`                        |

Arbitrary additional command bytestrings can also be generated as shown below.
This is a device-specific command for Brooks GF40 Mass Flow Controllers, which takes an IEE-754 floating point number as well as a unique code.

```python
import struct
import hart_protocol
code = 0
value = 32.1
data = struct.pack(">Bf", code, value)
command = hart_protocol.pack_command(address=123, command_id=236, data=data)
```

## Parsing Responses

All responses are parsed into named tuples.
Every single response will have the following keys.

Generic Response
| key             | value     |
| --------------- | --------- |
| `address`       | `<int>`   |
| `bytecount`     | `<int>`   |
| `command`       | `<int>`   |
| `command_name`  | `<str>`   |
| `data`          | `<bytes>` |
| `full_response` | `<bytes>` |
| `device_status` | `<bytes>` |
| `response_code` | `<bytes>` |

You can parse the raw `data` according to the particulars of your peripheral.
Certain standard responses are parsed further as shown below.

Response 0
| key                                           | value                      |
| --------------------------------------------- | -------------------------- |
| `command_name`                                | `"read_unique_identifier"` |
| `command`                                     | `0`                        |
| `device_id`                                   | `<bytes>`                  |
| `hardware_revision_level`                     | `<int>`                    |
| `manufacturer_device_type`                    | `<bytes>`                  |
| `manufacturer_id`                             | `<int>`                    |
| `number_response_preamble_charachters`        | `<int>`                    |
| `software_revision_level`                     | `<int>`                    |
| `transmitter_specific_command_revision_level` | `<int>`                    |
| `universal_command_revision_level`            | `<int>`                    |

Response 1
| key                | value                     |
| ------------------ | ------------------------- |
| `command_name`     | `"read_primary_variable"` |
| `command`          | `1`                       |
| `primary_variable` | `<float>`                 |

Response 11
| key                                           | value                      |
| --------------------------------------------- | -------------------------- |
| `command_name`                                | `"read_unique_identifier"` |
| `command`                                     | `11`                       |
| `device_id`                                   | `<bytes>`                  |
| `hardware_revision_level`                     | `<int>`                    |
| `manufacturer_device_type`                    | `<bytes>`                  |
| `manufacturer_id`                             | `<int>`                    |
| `number_response_preamble_charachters`        | `<int>`                    |
| `software_revision_level`                     | `<int>`                    |
| `transmitter_specific_command_revision_level` | `<int>`                    |
| `universal_command_revision_level`            | `<int>`                    |

Many other universal and common responses are also parsed..., give it a try!

## Integration Example

```python
>>> import hart_protocol
>>> import serial
>>>
>>> port = serial.Serial("/dev/ttyUSB0", 19200, timeout=0.1)
>>> port.parity = "O"
>>> port.stopbits = 1
>>> tag = hart_protocol.tools.pack_ascii("06C22300517"[-8:])
>>> port.write(hart_protocol.universal.read_unique_identifier_associated_with_tag(tag))
>>>
>>> unpacker = hart_protocol.Unpacker(port)
>>> for msg in unpacker:
...     print(msg)
...
>>>
```

## Maintainers

- [Blaise Thompson](https://github.com/untzag)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/).

## [Unreleased]

## [2023.6.0]

### Fixed
- VERSION is now a python file so that it can be recognized by pyinstaller

## [2023.5.0]

### Fixed
- missing final byte from message 16

## [2023.4.0]

### Fixed
- missing final byte from messages 12, 13, 14, 17, 18

## [2023.3.0]

### Changed
- in responses, device_status and response_code are now unpacked as two separate bytes
- data with bytecount of 2 has status only, and now returns immediately

## [2022.8.2]

### Fixed
- if the unpacker "ran ahead" of the serial buffer it would not succeed next time

## [2022.8.1]

### Fixed
- unpacker was broken in an async context

## [2022.8.0]

### Added
- initial release

[Unreleased]: https://github.com/yaq-project/hart-protocol/compare/v2023.6.0...main
[2023.6.0]: https://github.com/yaq-project/hart-protocol/compare/v2023.5.0...v2023.6.0
[2023.5.0]: https://github.com/yaq-project/hart-protocol/compare/v2023.4.0...v2023.5.0
[2023.4.0]: https://github.com/yaq-project/hart-protocol/compare/v2023.3.0...v2023.4.0
[2023.3.0]: https://github.com/yaq-project/hart-protocol/compare/v2022.8.2...v2023.3.0
[2022.8.2]: https://github.com/yaq-project/hart-protocol/compare/v2022.8.1...v2022.8.2
[2022.8.1]: https://github.com/yaq-project/hart-protocol/compare/v2022.8.0...v2022.8.1
[2022.8.0]: https://gitlab.com/yaq/yaqd-picotech/tags/v2022.8.0


hart_protocol/_parsing.py
import struct
from typing import MutableMapping, Union


def parse(response: bytes) -> MutableMapping[str, Union[int, bytes, str, float]]:
    out: MutableMapping[str, Union[int, bytes, str, float]] = dict()
    out["full_response"] = response
    if response[0] & 0x80:  # long address
        out["address"] = int.from_bytes(response[1:6], "big")
        response = response[6:]
    else:  # short address
        out["address"] = response[1]
        response = response[2:]
    command, bytecount, response_code, device_status = struct.unpack_from(">BBBB", response)
    out["device_status"] = device_status
    out["response_code"] = response_code
    data = response[4 : 4 + bytecount]
    out["command"] = command
    out["command_name"] = f"hart_command_{command}"
    out["bytecount"] = bytecount
    out["data"] = data

    # handle error return
    if bytecount == 2:
        return out

    # universal commands
    if command in [0, 11]:
        out["command_name"] = "read_unique_identifier"
        out["manufacturer_id"] = data[1]
        out["manufacturer_device_type"] = data[2]
        out["number_response_preamble_characters"] = data[3]
        out["universal_command_revision_level"] = data[4]
        out["transmitter_specific_command_revision_level"] = data[5]
        out["software_revision_level"] = data[6]
        out["hardware_revision_level"] = data[7]
        out["device_id"] = int.from_bytes(data[9:12], "big")
    elif command in [1]:
        out["command_name"] = "read_primary_variable"
        units, variable = struct.unpack_from(">Bf", data)
        out["primary_variable_units"] = units
        out["primary_variable"] = variable
    elif command in [2]:
        out["command_name"] = "read_loop_current_and_percent"
        analog_signal, primary_variable = struct.unpack_from(">ff", data)
        out["analog_signal"] = analog_signal
        out["primary_variable"] = primary_variable
    elif command in [3]:
        out["command_name"] = "read_dynamic_variables_and_loop_current"
        (
            analog_signal,
            primary_variable_units,
            primary_variable,
            secondary_variable_units,
            secondary_variable,
        ) = struct.unpack_from(">fBfBf", data)
        out["analog_signal"] = analog_signal
        out["primary_variable_units"] = primary_variable_units
        out["primary_variable"] = primary_variable
        out["secondary_variable_units"] = secondary_variable_units
        out["secondary_variable"] = secondary_variable
    elif command in [6]:
        out["command_name"] = "write_polling_address"
        polling_address = struct.unpack_from(">B", data)[0]
        out["polling_address"] = polling_address
    elif command in [12]:
        out["command_name"] = "read_message"
        out["message"] = data[0:24]
    elif command in [13]:
        out["command_name"] = "read_tag_descriptor_date"
        out["device_tag_name"] = data[0:6]
        out["device_descriptor"] = data[6:18]
        out["date"] = data[18:21]
    elif command in [14]:
        out["command_name"] = "read_primary_variable_information"
        out["serial_no"] = data[0:3]
        sensor_limits_code, upper_limit, lower_limit, min_span = struct.unpack_from(
            ">xxxBfff", data
        )
        out["sensor_limits_code"] = sensor_limits_code
        out["upper_limit"] = upper_limit
        out["lower_limit"] = lower_limit
        out["min_span"] = min_span
    elif command in [15]:
        out["command_name"] = "read_output_information"
        (
            alarm_code,
            transfer_fn_code,
            primary_variable_range_code,
            upper_range_value,
            lower_range_value,
            damping_value,
            write_protect,
            private_label,
        ) = struct.unpack_from(">BBBfffBB", data)
        out["alarm_code"] = alarm_code
        out["transfer_fn_code"] = transfer_fn_code
        out["primary_variable_range_code"] = primary_variable_range_code
        out["upper_range_value"] = upper_range_value
        out["lower_range_value"] = lower_range_value
        out["damping_value"] = damping_value
        out["write_protect"] = write_protect
        out["private_label"] = private_label
    elif command in [16]:
        out["command_name"] = "read_final_assembly_number"
        out["final_assembly_no"] = int.from_bytes(data[0:3], "big")
    elif command in [17]:
        out["command_name"] = "write_message"
        out["message"] = data[0:24]
    elif command in [18]:
        out["command_name"] = "write_tag_descriptor_date"
        out["device_tag_name"] = data[0:6]
        out["device_descriptor"] = data[6:18]
        out["date"] = data[18:21]
    elif command in [19]:
        out["command_name"] = "write_final_assembly_number"
        out["final_assembly_no"] = int.from_bytes(data[0:2], "big")

    # COMMON COMMANDS

    # elif command in [37]:
    #     out["command_name"] = "set_primary_variable_lower_range_value"
    #     out[""] =
    # request data bytes = NONE, response data bytes = NONE

    # elif command in [38]:
    #     out["command_name"] = "reset_configuration_changed_flag"
    #     out[""] =
    # request data bytes = NONE, response data bytes = NONE

    # elif command in [42]:
    #     out["command_name"] = "perform_master_reset"
    #     out[""] =
    # request data bytes = NONE, response data bytes = NONE

    # elif command in [48]:
    #     out["command_name"] = "read_additional_transmitter_status"
    #     out[""] =
    # request data bytes = NONE, response data bytes = NONE

    elif command in [50]:
        out["command_name"] = "read_dynamic_variable_assignments"
        (
            primary_transmitter_variable,
            secondary_transmitter_variable,
            tertiary_transmitter_variable,
            quaternary_transmitter_variable,
        ) = struct.unpack_from(">BBBB", data)
        out["primary_transmitter_variable"] = primary_transmitter_variable
        out["secondary_transmitter_variable"] = secondary_transmitter_variable
        out["tertiary_transmitter_variable"] = tertiary_transmitter_variable  # NOT USED
        out["quaternary_transmitter_variable"] = quaternary_transmitter_variable  # NOT USED
    elif command in [59]:
        out["command_name"] = "write_number_of_response_preambles"
        n_response_preambles = struct.unpack_from(">B", data)[0]
        out["n_response_preambles"] = n_response_preambles
    elif command in [66]:
        out["command_name"] = "toggle_analog_output_mode"
        (
            analog_output_selection,
            analog_output_units_code,
            fixed_analog_output,
        ) = struct.unpack_from(">BBf", data)
        out["analog_output_selection"] = analog_output_selection
        out["analog_output_units_code"] = analog_output_units_code
        out["fixed_analog_output"] = fixed_analog_output
    elif command in [67]:
        out["command_name"] = "trim_analog_output_zero"
        analog_output_code, analog_output_units_code, measured_analog_output = struct.unpack_from(
            ">BBf", data
        )
        out["analog_output_code"] = analog_output_code
        out["analog_output_units_code"] = analog_output_units_code
        out["measured_analog_output"] = measured_analog_output
    elif command in [68]:
        out["command_name"] = "trim_analog_output_span"
        analog_output_code, analog_output_units_code, measured_analog_output = struct.unpack_from(
            ">BBf", data
        )
        out["analog_output_code"] = analog_output_code
        out["analog_output_units_code"] = analog_output_units_code
        out["measured_analog_output"] = measured_analog_output
    elif command in [123]:
        out["command_name"] = "select_baud_rate"
        out["baud_rate"] = int.from_bytes(data, "big")

    return out
